x/dualstaking/keeper/delegate.go

```diff
@@ -0,0 +1,572 @@
+			utils.Attribute{Key: "provider", Value: provider},
+		)
+	}
+
+	if err := validateCoins(amount); err != nil {
+		return err
+	} else if amount.IsZero() {
+		return nil
+	}
+
+	err = k.decreaseDelegation(ctx, delegator, provider, chainID, amount, nextEpoch)
+	if err != nil {
+		return utils.LavaFormatWarning("failed to decrease delegation", err,
+			utils.Attribute{Key: "delegator", Value: delegator},
+			utils.Attribute{Key: "provider", Value: provider},
+			utils.Attribute{Key: "amount", Value: amount.String()},
+		)
+	}
+
+	// in unbonding the funds to not return immediately to the delegator; instead
+	// they transfer to the BonderPoolName module account, where they are held for
+	// the hold period before they are finally released.
+
+	err = k.bankKeeper.SendCoinsFromModuleToModule(ctx, types.NotBondedPoolName, types.BondedPoolName, sdk.NewCoins(amount))
+	if err != nil {
+		return utils.LavaFormatError("failed to unbond delegated coins", err,
+			utils.Attribute{Key: "amount", Value: amount},
+		)
+	}
+
+	err = k.setUnbondingTimer(ctx, delegator, provider, chainID, amount)
+	if err != nil {
+		return utils.LavaFormatError("failed to unbond delegated coins", err,
+			utils.Attribute{Key: "amount", Value: amount},
+		)
+	}
+
+	details := map[string]string{
+		"delegator": delegator,
+		"provider":  provider,
+		"chainID":   chainID,
+		"amount":    amount.String(),
+	}
+
+	utils.LogLavaEvent(ctx, k.Logger(ctx), types.UnbondingEventName, details, "Unbond")
+
+	return nil
+}
+
+var space = " "
+
+// encodeForTimer generates timer key unique to a specific delegation; thus it
+// must include the delegator, provider, and chainID.
+func encodeForTimer(delegator, provider, chainID string) []byte {
+	dlen, plen, clen := len(delegator), len(provider), len(chainID)
+
+	encodedKey := make([]byte, dlen+plen+clen+2)
+
+	index := 0
+	index += copy(encodedKey[index:], delegator)
+	encodedKey[index] = []byte(space)[0]
+	index += copy(encodedKey[index+1:], provider) + 1
+	encodedKey[index] = []byte(space)[0]
+	copy(encodedKey[index+1:], chainID)
+
+	return encodedKey
+}
+
+// decodeForTimer extracts the delegation details from a timer key.
+func decodeForTimer(encodedKey []byte) (string, string, string) {
+	split := strings.Split(string(encodedKey), space)
+	if len(split) != 3 {
+		return "", "", ""
+	}
+	delegator, provider, chainID := split[0], split[1], split[2]
+	return delegator, provider, chainID
+}
+
+// setUnbondingTimer creates a timer for an unbonding operation, that will expire
+// in the spec's unbonding-hold-blocks blocks from now. only one unbonding request
+// can be placed per block (for a given delegator/provider/chainID combination).
+func (k Keeper) setUnbondingTimer(ctx sdk.Context, delegator, provider, chainID string, amount sdk.Coin) error {
+	key := encodeForTimer(delegator, provider, chainID)
+
+	period := k.stakingKeeper.UnbondingTime(ctx)
+	timeout := ctx.BlockTime().Add(period).UTC().Unix()
```

Comment:
do this for unstakeholdblocks instead of time.
this will be changed once we align the unstake of regulare provider, we need those to be the same

Reply:
discussed with Omer who recommended to use time based hold-period like cosmos's `x/staking`, and specifically use the same period from that module.

Reply:
changed in 5c318faa